### ISSUE TITLE
Timeout: Cannot save running-config with nxos_command or save: yes in nxos_config

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29592]
From @jean-christophe-manciot on 2016-10-10T18:08:05Z
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

nxos_command
nxos_config
##### ANSIBLE VERSION

```
ansible 2.3.0~git20161010.03765ba
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

inventory   = ./hosts
gathering = explicit
roles_path = /home/actionmystique/Program-Files/Ubuntu/Ansible/Roles/roles
private_role_vars = yes
log_path = /var/log/ansible.log
fact_caching = redis
fact_caching_timeout = 86400
retry_files_enabled = False
##### OS / ENVIRONMENT
- host: Ubuntu 16.04 4.4.0
- target: **NX-OSv 7.3(0)D1(1)**
##### SUMMARY

cf. title
##### STEPS TO REPRODUCE

**Inventory ./hosts**:

```
[all:vars]
nms_mgt_ip_address=172.21.100.1
[spines]
NX_OSv_Spine_11 ansible_host=172.21.100.11 
NX_OSv_Spine_12 ansible_host=172.21.100.12 
```

Structure passed as "**provider**": connections.ssh
Defined in group_vars/nx_osv/connections.yml and a symbolic link in roles/nxos_snmp/defaults points to nx_osv

```
connections
...
        nxapi:
          transport: nxapi
          host: "{{ ansible_host }}"
          # ansible_port
          port: "{{ http.port }}"
          # ansible_user
          username: admin
          password: xxxxxxxx
          # enable_secret_password
          auth_pass: xxxxxxxx
          # http or https
          use_ssl: no
          validate_certs: "{{ https.validate_certs }}"
```

**Role**: nxos_snmp:

```
- include_vars: "../defaults/{{ os_family }}/connections.yml"
...
- name: Saving the running-config
  nxos_command:
        provider: "{{ connections.nxapi }}"
        commands:
          - "copy running-config startup-config"
  register: result
```

**Playbook**:

```
- name: Configuring SNMP on NX-OS/NX-OSv
  hosts:
    - nx_osv
  roles:
    - nxos_snmp
```
##### EXPECTED RESULTS

We should be able to save the modifications.
##### ACTUAL RESULTS

```
TASK [nxos_snmp : Saving the running-config] ***********************************
fatal: [NX_OSv_Spine_12]: FAILED! => {"changed": false, "clierror": "Syntax error while parsing 'copy running-config startup-config  | xml '\n\n\nCmd exec error.\n", "code": "400", "failed": true, "input": "copy running-config startup-config", "msg": "Input CLI command error", "output": {"clierror": "Syntax error while parsing 'copy running-config startup-config  | xml '\n\n\nCmd exec error.\n", "code": "400", "input": "copy running-config startup-config", "msg": "Input CLI command error"}, "url": "http://172.21.100.12:8080/ins"}
```

No issue when configuring through the CLI:

```
NX_OSv_Spine_12# copy running-config startup-config 
[########################################] 100%
Copy complete.
NX_OSv_Spine_12# show startup-config

!Command: show startup-config
!Time: Mon Oct 10 18:12:08 2016
!Startup config saved at: Mon Oct 10 18:01:55 2016

version 7.3(0)D1(1)
power redundancy-mode redundant

```

Copied from original issue: ansible/ansible-modules-core#5215

**Comment by ansibot** [comment]
From @jean-christophe-manciot on 2016-10-10T18:08:05Z

Using `save: yes` in `nxos_config` leads to no error during execution, but afterwards:

```
NX_OSv_Spine_11# show startup-config 
2016 Oct 10 18:11:45 NX_OSv_Spine_11 %PSS-2-PSS_INVALID_SIZE: ngoam: database /var/sysmgr/startup-cfg/bin/ngoam_config (.pss) has invalid size 0 (block size -1)
```

**Comment by ansibot** [comment]
From @ansibot on 2016-10-10T18:08:05Z

@jedelman8, @GGabriele, @privateip, @gundalow, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @jean-christophe-manciot on 2016-10-10T18:08:05Z

A workaround is to replace nxos_command module call by nxos_config.

**Comment by ansibot** [comment]
From @jean-christophe-manciot on 2016-10-10T18:08:05Z

The previous workaround is not possible anymore with ansible 2.3.0~git20161011.6be09ee:

```
- name: Saving the running-config
  nxos_config:
        provider: "{{ connections.nxapi }}"
        commands:
          - "copy running-config startup-config"
  register: result
```

leads to:

```
TASK [nxos_snmp : Saving the running-config] ***********************************
fatal: [NX_OSv_Spine_12]: FAILED! => {"changed": false, "clierror": "ERROR: Copy r s / partial mem write in progress. Please try after some time.\n", "code": "400", "failed": true, "input": "rollback running-config checkpoint ansible_1476288874", "msg": "CLI execution error", "output": {"clierror": "ERROR: Copy r s / partial mem write in progress. Please try after some time.\n", "code": "400", "input": "rollback running-config checkpoint ansible_1476288874", "msg": "CLI execution error"}, "url": "http://172.21.100.12:8080/ins"}
```

It is barely possible to perform the same task from the CLI:

```
copy running-config startup-config 
[#################                       ]  42%2016 Oct 12 16:26:15 NX_OSv_Spine_12 %PSS-2-PSS_ERROR: Failed to add pages to director, PSS_MALLOC failed, errno 12 (Cannot allocate memory)
[##################                      ]  43%2016 Oct 12 16:26:15 NX_OSv_Spine_12 %PSS-2-PSS_ERROR: Page director has non zero number of pages
[########################################]  100%
Copy complete.
```

It took a time too long for ansible to complete.
However, I suspect that the multiple nxos_config which have run previously are responsible for this situation, because throwing more RAM does not remove the error.

**Comment by ansibot** [comment]
From @jean-christophe-manciot on 2016-10-10T18:08:05Z

With ansible 2.2.0.0-0.1-rc1, running exactly the same set of nxos_config followed by a save of the running-config is successful with the CLI:

```
NX_OSv_Spine_12# copy running-config startup-config 
[########################################] 100%
Copy complete.
```

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (120/100 (120%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext2.iad.github.net ([192.30.252.193]:37477 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ2D-000oBd-7n
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:50:01 -0500
Date: Mon, 11 Sep 2017 07:49:02 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141360;
	bh=jSwV6JUk+A7GNcqfWwTKt8WA4jnkfCQNEVXSxfrd9Zs=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=fxv/H5Gj3BVGcElp94MZQgcISW1lzPlK+XlO85PraQbuHxHKjEGOY08k8sk2sfiiu
	 s8Hdo62gc39pzvMzXMAT/7IY4poXfdwkjiIf+tcdXEt1kqk/TVa8rP3oIlkb1GqWOy
	 QCXh/wy2GXJMaEJtMbPYc9R3O4eW/FTLtE829RkY=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (121/100 (121%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 7710 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext4.iad.github.net ([192.30.252.195]:38363 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ2Y-000oMO-3W
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:50:22 -0500
Date: Mon, 11 Sep 2017 07:49:06 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141374;
	bh=BwSR+Q9HDnVBGn5rxR7OGleWDEdUzcY8uQi+o1FXr/M=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=iVyWoRHZgLjuBjuh8lxDV7uWzqB7m6haSVk7rWN/R4HXaBaZs2OaGOhDxmXx9M31e
	 V7XnH1B+00MfggfR+F/EpWzHJpkR+tER0YVAM0Xpbwb68vZKWQ5ZnIWcWb4fBrMvgN
	 uvuaTO/OrLnbAw+z5cmxJnzaT8BZwtihEfqtME/c=

**Comment by ansibot** [comment]
cc @ganeshrn @mikewiebe @rahushen @rcarrillocruz @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @kedarX
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
@jean-christophe-manciot there were a huge amount of NXOS bug fixes added in Ansible 2.4. We've also added a load of integration tests to defend this functionality.

Could you please retest in 2.4.

needs_info

**Comment by jean-christophe-manciot** [comment]
1) **The issue with nxos_command is still there** with:
```
NX-OSv-9k 7.0.3.I6.1

ansible 2.4.1.0 (detached HEAD 4a4590a2c6) last updated 2017/10/24 09:41:27 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  ansible python module location = /home/actionmystique/src/Ansible/git-ansible/lib/ansible
  executable location = /home/actionmystique/src/Ansible/git-ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```
Role:
```
- name: Listing configuration with NXAPI
  nxos_command:
        provider: "{{ connections.nxapi }}"
        commands:
          - "show running-config"
  register: result

- name: Saving configuration with SSH
  nxos_command:
        provider: "{{ connections.ssh }}"
        commands:
          - "copy running-config startup-config"
  register: result

- name: Saving configuration with NXAPI
  nxos_command:
        provider: "{{ connections.nxapi }}"
        commands:
          - "copy running-config startup-config"
  register: result
```
The first play is successful, the second is **successful with SSH transport**, the last is **unsuccessful with NXAPI transport**:
```
< TASK [nxos_yang : Saving configuration with NXAPI] >
 -----------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: /home/actionmystique/Ansible/Roles/roles/nxos_yang/tasks/init.yml:71
<172.21.100.18> connection transport is nxapi
Using module file /home/actionmystique/src/Ansible/git-ansible/lib/ansible/modules/network/nxos/nxos_command.py
<172.21.100.18> ESTABLISH LOCAL CONNECTION FOR USER: root
<172.21.100.18> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
fatal: [NX_OSv_9k_Spine_31]: FAILED! => {
    "changed": false, 
    "clierror": "Syntax error while parsing 'copy running-config startup-config  | xml '\n\n\nCmd exec error.\n", 
    "code": "400", 
    "failed": true, 
    "input": "copy running-config startup-config", 
    "invocation": {
        "module_args": {
            "commands": [
                "copy running-config startup-config"
            ], 
            "host": "172.21.100.18", 
            "interval": 1, 
            "match": "all", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 80, 
            "provider": {
                "host": "172.21.100.18", 
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "port": 80, 
                "ssh_keyfile": null, 
                "timeout": 60, 
                "transport": "nxapi", 
                "use_ssl": false, 
                "username": "admin", 
                "validate_certs": false
            }, 
            "retries": 10, 
            "ssh_keyfile": null, 
            "timeout": 60, 
            "transport": "nxapi", 
            "url_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "url_username": "admin", 
            "use_ssl": false, 
            "username": "admin", 
            "validate_certs": false, 
            "wait_for": null
        }
    }, 
    "msg": "Input CLI command error", 
    "output": {
        "clierror": "Syntax error while parsing 'copy running-config startup-config  | xml '\n\n\nCmd exec error.\n", 
        "code": "400", 
        "input": "copy running-config startup-config", 
        "msg": "Input CLI command error"
    }, 
    "url": "http://172.21.100.18:80/ins"
}
```

Exactly the same results with the latest unstable version:
```
ansible 2.5.0 (devel 9d6fd4ebcc) last updated 2017/10/24 10:19:35 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  ansible python module location = /home/actionmystique/src/Ansible/git-ansible/lib/ansible
  executable location = /home/actionmystique/src/Ansible/git-ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

**Comment by jean-christophe-manciot** [comment]
2) **The issue with nxos_config/save_when is solved** with both previous ansible releases from stable-2.4 & devel branches.

**Comment by ansibot** [comment]
@ansibot This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
Sounds like this is fixed, so closing
